### PR TITLE
[Return-types #12.2] JAX-JIT interface rework with QNode integration: device gradients forward mode

### PR DESCRIPTION
**Pull request by antalszava**
**Context:**
Same as for https://github.com/PennyLaneAI/pennylane/pull/3235.

This PR adds the logic for `mode="forward"` for `diff_method="device"`.

### CODE REVIEW

File: pennylane/interfaces/jax_jit_tuple.py [hunk 1]
```diff
@@ -296,4 +296,55 @@ def _execute_fwd_tuple(
     gradient_kwargs=None,
     _n=1,
 ):  # pylint: disable=dangerous-default-value,unused-argument
-    raise NotImplementedError("Forward mode execution for device gradients is not yet implemented.")
+    """The auxiliary execute function for cases when the user requested
+    jacobians to be computed in forward mode (e.g. adjoint) or when no gradient function was
+    provided. This function does not allow multiple derivatives. It currently does not support shot vectors
+    because adjoint jacobian for default qubit does not support it.."""
```
Comment (rmoyard) [inline pennylane/interfaces/jax_jit_tuple.py:302]:
```suggestion
    because adjoint jacobian for default qubit does not support it."""
```

**Review by rmoyard — APPROVED** [review]
Thanks @antalszava for that PR! It great to have the jitting interfaces to work for multiple measurements with first derivative 🥇

File: doc/releases/changelog-dev.md [hunk 2]
```diff
@@ -417,6 +417,7 @@
 * The JAX-JIT interface now supports gradient transforms and device gradient execution in `backward` mode with the new
   return types system.
   [(#3235)](https://github.com/PennyLaneAI/pennylane/pull/3235)
+  [(#3445)](https://github.com/PennyLaneAI/pennylane/pull/3445)
```
Comment (eddddddy) [inline doc/releases/changelog-dev.md:420]: I think this changelog entry should mention that forward mode is also supported now.

File: pennylane/interfaces/jax_jit_tuple.py [hunk 3]
```diff
@@ -296,4 +296,55 @@ def _execute_fwd_tuple(
...
+        return res, jvps
+
+    res = execute_wrapper(params)
+
+    tracing = []
+    for i, tape in enumerate(tapes):
+        if len(tape.measurements) == 1:
+            tracing.append(isinstance(res[i], jax.interpreters.ad.JVPTracer))
+        else:
+            tracing.extend([isinstance(r, jax.interpreters.ad.JVPTracer) for r in res[i]])
```
Comment (eddddddy) [inline pennylane/interfaces/jax_jit_tuple.py:340]: Is it required to iterate over the results to determine if we are tracing, instead of just looking at the first one? Are the scenarios where some results are tracers but others are not?

**Review by eddddddy — APPROVED** [review]
Clean and concise @antalszava! Just have one question.

File: doc/releases/changelog-dev.md (hunk 2 above)
Comment (antalszava) [reply doc/releases/changelog-dev.md]: Good catch! I've removed the specifics mentioned as now all the cases other interfaces support are also supported here.

File: pennylane/interfaces/jax_jit_tuple.py (hunk 3 above)
Comment (antalszava) [reply pennylane/interfaces/jax_jit_tuple.py:340]: Good question! I don't think so actually :thinking: I'm changing the list comprehension here to using `any` with a generator - that case would cover either case, especially so because we're checking the overall `tracing` sequence using `any`.

I've looked at some cases now, but the boolean seem to have matched for multiple measurements.

**Comment by Qottmann** [comment]
Curiosity question: Is there currently a user-facing (or in any way accessible) way to enforce forward mode differentiation in qnodes? Is this planned to be a feature or is this just required for efficient higher order derivatives?

**Comment by PhilipVinc** [comment]
low level, `jax.jvp` is forward mode and `jax.vjp` is backward mode.
You can also just use `jax.jacfwd` and `jax.jacrev`.

**Comment by antalszava** [comment]
Exactly. Also, the naming atm is a bit confusing: the `mode` keyword argument passed to a QNode applies to device gradients of computing quantum gradients. It doesn't turn on forward mode automatic differentiation but rather computes the gradient in the function that is evaluated by the ML framework on the forward pass (backward mode does the same on the backward pass).
